src/modules/roc_audio/profiler.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include "roc_audio/profiler.h"
+#include "roc_core/log.h"
+#include "roc_core/panic.h"
+#include "roc_core/shared_ptr.h"
+
+namespace {
+const size_t chunk_time = 10;
+}
+
+namespace roc {
+namespace audio {
+
+Profiler::Profiler(core::IAllocator& allocator,
+                   packet::channel_mask_t channels,
+                   size_t sample_rate,
+                   core::nanoseconds_t interval)
+    : rate_limiter_(interval)
+    , interval_(interval)
+    , chunk_length_(chunk_time * sample_rate / (core::Second / core::Millisecond))
+    , num_chunks_((size_t)((unsigned long)interval
+                           / (unsigned long)(chunk_time * core::Millisecond))
+                  + 1)
+    , chunks_(allocator)
+    , first_chunk_num_(0)
+    , last_chunk_num_(0)
+    , last_chunk_samples_(0)
+    , moving_avg_(0)
+    , sample_rate_(sample_rate)
+    , num_channels_(packet::num_channels(channels)) {
+    if (num_channels_ == 0) {
+        roc_panic("profiler: n_channels is zero");
+    }
+    if (sample_rate_ == 0) {
+        roc_panic("profiler: sample_rate is zero");
+    }
+
+    if (!chunks_.resize(num_chunks_)) {
+        roc_log(LogError, "profiler: can't allocate chunks");
+        return;
+    }
+
+    valid_ = true;
+}
+
+bool Profiler::valid() const {
+    return valid_;
+}
+
+void Profiler::add_frame(size_t frame_size, core::nanoseconds_t elapsed) {
+    roc_panic_if(!valid_);
+
+    if (frame_size % num_channels_ != 0) {
+        roc_panic("profiler: unexpected frame size");
+    }
+
+    const double frame_speed =
```

Comment:
Nitpick: maybe it makes sense to extract the calculation (lines 64-101) into a private method, say, update_avg_().